### ISSUE TITLE
Issue when passing a union typedesc of custom types

### ISSUE THREAD

**Issue by irshadnilam** [ballerina-platform/ballerina-lang#23127]
**Description:**
Please consider following code snippet

```ballerina
type Person record {
    string name;
    int age;
};

type Employee record {
    string name;
    int age;
    int empNo;
};

public function main() {
    test(Person|Employee); // this produces a compile time error

    // the following works
    test(string|int);
    test(typedesc<Person|Employee>);
}

public function test(typedesc<any> a) {
}
```
The above shows a compilation error. 
`operator '|' not defined for 'typedesc<Person>' and 'typedesc<Employee>'`

**Comment by irshadnilam** [comment]
I'll open a spec issue to get some clarification on this. I'll link the spec-issue once it is opened.

**Comment by MaryamZi** [comment]
I believe the conclusion on this was that the error is correct.

However, the compiler seems to crash for this sample on sl-alpha.

```log
[2021-01-26 23:56:35,950] SEVERE {b7a.log.crash} - class org.wso2.ballerinalang.compiler.tree.types.BLangBuiltInRefTypeNode cannot be cast to class org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression (org.wso2.ballerinalang.compiler.tree.types.BLangBuiltInRefTypeNode and org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression are in unnamed module of loader 'app') 
java.lang.ClassCastException: class org.wso2.ballerinalang.compiler.tree.types.BLangBuiltInRefTypeNode cannot be cast to class org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression (org.wso2.ballerinalang.compiler.tree.types.BLangBuiltInRefTypeNode and org.wso2.ballerinalang.compiler.tree.expressions.BLangExpression are in unnamed module of loader 'app')
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4663)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1928)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.BinaryExpressionNode.apply(BinaryExpressionNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:4692)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4663)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1928)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.BinaryExpressionNode.apply(BinaryExpressionNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:4692)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4663)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2951)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.PositionalArgumentNode.apply(PositionalArgumentNode.java:46)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:4692)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4663)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.lambda$createBLangInvocation$14(BLangNodeTransformer.java:5130)
	at java.base/java.util.Iterator.forEachRemaining(Iterator.java:133)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createBLangInvocation(BLangNodeTransformer.java:5130)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2001)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.FunctionCallExpressionNode.apply(FunctionCallExpressionNode.java:58)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createActionOrExpression(BLangNodeTransformer.java:4692)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createExpression(BLangNodeTransformer.java:4663)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:2835)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.ExpressionStatementNode.apply(ExpressionStatementNode.java:50)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.generateAndAddBLangStatements(BLangNodeTransformer.java:4572)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1483)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.FunctionBodyBlockNode.apply(FunctionBodyBlockNode.java:59)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.populateFunctionNode(BLangNodeTransformer.java:1391)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.createFunctionNode(BLangNodeTransformer.java:1371)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:1349)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.FunctionDefinitionNode.apply(FunctionDefinitionNode.java:71)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:569)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.transform(BLangNodeTransformer.java:462)
	at io.ballerina.compiler.syntax.tree.ModulePartNode.apply(ModulePartNode.java:54)
	at org.wso2.ballerinalang.compiler.parser.BLangNodeTransformer.accept(BLangNodeTransformer.java:492)
	at io.ballerina.projects.DocumentContext.compilationUnit(DocumentContext.java:111)
	at io.ballerina.projects.ModuleContext.compileInternal(ModuleContext.java:355)
	at io.ballerina.projects.ModuleCompilationState$1.compile(ModuleCompilationState.java:45)
	at io.ballerina.projects.ModuleCompilationState$1.generatePlatformSpecificCode(ModuleCompilationState.java:53)
	at io.ballerina.projects.ModuleContext.generatePlatformSpecificCode(ModuleContext.java:326)
	at io.ballerina.projects.JBallerinaBackend.performCodeGen(JBallerinaBackend.java:136)
	at io.ballerina.projects.JBallerinaBackend.<init>(JBallerinaBackend.java:122)
	at io.ballerina.projects.JBallerinaBackend.lambda$from$0(JBallerinaBackend.java:98)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1133)
	at io.ballerina.projects.PackageCompilation.getCompilerBackend(PackageCompilation.java:135)
	at io.ballerina.projects.JBallerinaBackend.from(JBallerinaBackend.java:97)
	at io.ballerina.cli.task.CompileTask.execute(CompileTask.java:69)
	at io.ballerina.cli.TaskExecutor.executeTasks(TaskExecutor.java:40)
	at io.ballerina.cli.cmd.RunCommand.execute(RunCommand.java:155)
	at java.base/java.util.Optional.ifPresent(Optional.java:183)
	at io.ballerina.cli.launcher.Main.main(Main.java:58)
 ```

**Comment by lochana-chathura** [comment]
> However, the compiler seems to crash for this sample on sl-alpha.

Crashing fixed in #29874.

**Comment by MaryamZi** [comment]
Closing this since the crash is fixed and the code should (and does) result in compilation errors atm.

Related spec issue: https://github.com/ballerina-platform/ballerina-spec/issues/797
